### PR TITLE
Phix remover

### PR DESCRIPTION
**Pull request by dstreett**
Hey, @joe-angell . Mind checking out this branch?

I will switch the lookup table from a size_t to a bitset if you think it will look better.

### CODE REVIEW

File: Phix-Remover/src/phix_remover.h [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    else if (c == 'C')
+        return 1;
+    else if (c == 'G')
+        return 2;
+    return 5;
+}
+
+size_t check_read(const Read &r, std::array<size_t, 1<<(2*8) > lookup, std::array< size_t, 1 << (2*8) > lookup_rc, size_t kmerSize) {
+    size_t index = 0;
+    size_t kmer = kmerSize - 1;
```
Comment (joe-angell) [inline Phix-Remover/src/phix_remover.h]: if kmerSize == 0, this will be a problem.  assert(kmerSize > 0)

File: Phix-Remover/src/phix_remover.h [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    else if (c == 'T')
+        return 3;
+    else if (c == 'C')
+        return 1;
+    else if (c == 'G')
+        return 2;
+    return 5;
+}
+
+size_t check_read(const Read &r, std::array<size_t, 1<<(2*8) > lookup, std::array< size_t, 1 << (2*8) > lookup_rc, size_t kmerSize) {
```
Comment (joe-angell) [inline Phix-Remover/src/phix_remover.h]: why 1<<(2*8)?  Maybe add that as a const with some description as to it's  purpose

File: Phix-Remover/src/phix_remover.h [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    std::string seq = r.get_seq();
+    
+    for (std::string::iterator bp = seq.begin(); bp < seq.end(); ++bp) {
+        
+        if ((bin = getBin(*bp)) == 5) { //bp = N
+            kmer = kmerSize - 1;
+            index = 0;
+        }
+
+        index &= ~(3 << 14); //only check forward strand against both RC and non-rc
```
Comment (joe-angell) [inline Phix-Remover/src/phix_remover.h]: describe intent, can't tell if this is correct

File: Phix-Remover/src/phix_remover.cpp [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+                }
+                out_1.reset(new HtsOfstream(default_outfiles[0], force, gzip_out, std_out));
+
+                pe.reset(new ReadBaseOutTab(out_1));
+                se.reset(new ReadBaseOutTab(out_1));
+            }
+
+            Read readPhix = Read(phix, "", "");
+           
+            std::array<size_t, 1<< (2*kmer)>  lookup;
```
Comment (joe-angell) [inline Phix-Remover/src/phix_remover.cpp]: is there a reason you chose array over vector for all these?   I would definitely typedef this so you don't have to repeat it everywhere.

File: Phix-Remover/src/phix_remover.h (hunk 2 above)
Comment (joe-angell) [inline Phix-Remover/src/phix_remover.h]: I think you want lookup and lookup_rc to be passed by reference here?

**Comment by dstreett** [comment]
Hey, @joe-angell . I'll give you a little bit more of a description, sorry about that!

The idea is we break up the small PhiX genome into little 8-mers (overlapping 8 bp sequences). We then use these 8-mers as "lookups". If the sequence in question has enough hits to this lookup, we consider it Phix and it is not written out, otherwise, it is written out not changed.

I went with the binary representation of the sequence because I can have a constant lookup time versus N log N with a map of some sort. Additionally, an array with all 8mers possible is only 65k elements, which isn't too bad to keep in memory.

Great suggestions, I will make those changes and push it back up.

**Comment by dstreett** [comment]
Let me know if this is looking better.

File: Phix-Remover/src/phix_remover.h [hunk 5]
```diff
@@ -27,49 +28,51 @@ uint8_t getBin(char c) {
     return 5;
 }
 
-size_t check_read(const Read &r, kmerArray lookup, kmerArray lookup_rc, size_t kmerSize) {
-    size_t index = 0;
-    size_t kmer = kmerSize - 1;
+size_t check_read(const Read &r, kmerArray lookup, kmerArray lookup_rc) {
```
Comment (joe-angell) [inline Phix-Remover/src/phix_remover.h]: I'd like to see more documentation on what this function is supposed to do, just a comment at the top or something.

File: Phix-Remover/src/phix_remover.h [hunk 6]
```diff
@@ -93,40 +96,53 @@ void helper_discard(InputReader<T, Impl> &reader, std::shared_ptr<OutputWriter>
 }
 
 /*Hand phix read to set lookup tables*/
-void setLookup(kmerArray &lookup, kmerArray &lookup_rc, Read &rb, size_t kmerSize) {
-    size_t index = 0;
-    size_t index_rc = 0;
-    size_t kmer = kmerSize - 1;
+void setLookup(kmerArray &lookup, kmerArray &lookup_rc, Read &rb) {
```
Comment (joe-angell) [inline Phix-Remover/src/phix_remover.h]: same with this function

File: Phix-Remover/src/phix_remover.h [hunk 7]
```diff
@@ -93,40 +96,53 @@ void helper_discard(InputReader<T, Impl> &reader, std::shared_ptr<OutputWriter>
...
-        if ((bin = getBin(*bp)) == 5) { //bp = N
-            kmer = kmerSize - 1;
-            index = 0;
-            index_rc = 0;
+        /*Cannot handle Ns in 2 bit form
+         * reset kmerCheck and bits*/
+        if ((bin = getBin(*bp)) == 5) { 
+            kmerCheck = kmer - 1;
+            forwardBits.reset();
+            reverseBits.reset();
```
Comment (joe-angell) [inline Phix-Remover/src/phix_remover.h]: do you need a continue here?

File: Phix-Remover/src/phix_remover.h (hunk 5 above)
Comment (dstreett) [inline Phix-Remover/src/phix_remover.h]: So, the lookup tables should not be changed at that point. They should be set and remain constant for all reads. Should I still pass by ref then?

File: Phix-Remover/src/phix_remover.h [hunk 8]
```diff
@@ -13,6 +13,7 @@
 
 typedef std::unordered_map<std::string, size_t> Counter;
 const size_t kmer = 8;
+const size_t kmerBits = kmer*2;
 typedef std::array< size_t, 1<<(2*kmer) > kmerArray;
```
Comment (joe-angell) [inline Phix-Remover/src/phix_remover.h]: I'm saying:  `typedef std::array< unit_16_t, 1<<(2*kmer)> kmerArray;`
isn't the data you are storing in the array 16bits?  Maybe i'm missing something, but that would quarter the memory usage.

File: Phix-Remover/src/phix_remover.h (hunk 5 above)
Comment (joe-angell) [inline Phix-Remover/src/phix_remover.h]: pass by const ref then, making a copy of the whole array is not very efficient.

Comment (dstreett) [inline Phix-Remover/src/phix_remover.h]: Ahhh, yeah, sorry, didn't think about that. Thanks!!

**Pull request merged**